service-base/src/main/java/org/eclipse/hono/service/credentials/validators/CredentialsValidatorHashedPassword.java

```diff
@@ -0,0 +1,98 @@
+
+    /**
+     * Get the type of credentials secrets this validator is responsible for.
+     *
+     * @return The type of credentials secrets.
+     */
+    @Override
+    public String getSecretsType() {
+        return CredentialsConstants.SECRETS_TYPE_HASHED_PASSWORD;
+    }
+
+    /**
+     * Approve a single secret (as JsonObject) against a given password.
+     * If the password matches against it, the approval is considered successful.
+     * <p>This involves the hash-function found in the credentials secrets record to hash the given password before the
+     * comparison against the password in the credentials secrets record is done.
+     * <p>If a salt is contained in the credentials record (in base64 encoding), the hash function will be salted first.
+     *
+     * @param password The password to validate.
+     * @param aSecret The secret record as JsonObject (as returned by the <a href="https://www.eclipse.org/hono/api/Credentials-API/">Credentials API</a>.
+     * @return The result of the approval as boolean.
+     */
+    @Override
+    protected boolean validateSingleSecret(final String password, final JsonObject aSecret) {
+        String hashFunction = aSecret.getString(CredentialsConstants.FIELD_SECRETS_HASH_FUNCTION);
+        if (hashFunction == null) {
+            return false;
+        }
+
+        return checkPasswordAgainstSecret(aSecret, hashFunction, password);
+    }
+
+    private boolean checkPasswordAgainstSecret(final JsonObject secret, final String hashFunction, final String protocolAdapterPassword) {
+
+        String pwdHash = secret.getString(CredentialsConstants.FIELD_SECRETS_PWD_HASH);
+        if (pwdHash == null) {
+            return false;
+        }
+
+        byte[] password = Base64.getDecoder().decode(pwdHash);
+
+        // TODO: performance - always decode salts is terrible
+        final String salt = secret.getString(CredentialsConstants.FIELD_SECRETS_SALT);
+        final byte[] decodedSalt = Base64.getDecoder().decode(salt);
+        final String decodedSaltStr = new String(decodedSalt);
+
+        try {
+            byte[] hashedPassword = hashPassword(hashFunction, decodedSaltStr, protocolAdapterPassword);
+            if (!Arrays.equals(password, hashedPassword)) {
+                return false;
+            }
+        } catch (NoSuchAlgorithmException e) {
+            return false;
+        } catch (UnsupportedEncodingException e) {
+            return false;
+        }
+        // check if the password is the hashed version of the protocol adapter password
+        return true;
+    }
+
+    private byte[] hashPassword(final String hashFunction,final String hashSalt, final String passwordToHash) throws NoSuchAlgorithmException, UnsupportedEncodingException {
+        MessageDigest messageDigest = MessageDigest.getInstance(hashFunction);
+        messageDigest.update(hashSalt.getBytes("UTF-8"));
```

Comment:
see? you immediately translate the salt back to bytes anyway ... and, BTW, using UTF-8 here whereas you are using ISO-8859-1 above on Windows ...